### ISSUE TITLE
modularity

### ISSUE THREAD

**Issue by tonymorris** [scalaz/scalaz#895]
Scalaz 8 has been discussed for a long time. Now with the Scalaz project reverted to being motivated by its original progressive goals, some of the long-standing issues could be addressed with an incompatible implementation of the library.

Two issues that stand out for Scalaz:

1. The use of subtyping.

2. The way by which modules are split throughout the project.

To a naive observer, these appear to be independent of each other. Beware.

I think it can be concluded that subtyping in Scala is a failed experiment. The
use of subtyping, even among type-classes, leads to problems, which I hope to 
discuss further in this issue. The downside to this is a code blowout. For example, the `Monad` type-class with *have a* `Functor` instance, and therefore, also a `fmap` method. This will simply delegate, leading type-class authors to write this boilerplate for each introduced type-class and growing as the relationships grow from the premise. I think this is a **much** smaller penalty to incur than the one which Scalaz currently incurs.

As for splitting out modules, a first approximation might be to split into data-types and type-classes. However, this raises an obvious question; what about data-types that depend on type-classes (and what about type-classes that depend on data-types)? Also, where do the instances for standard data types belong? Does `Functor[List]` belong in `Functor.scala` or `List.scala` or somewhere else? What about `Functor[scalaz.IList]`? It is important that instances are not orphaned.

One resolution (the current one), is to concede the point — data-types and type-classes are mish-mashed together and depend all among each other. However, this is biting off a bit too much. There exist type-classes that have no data-type dependencies, and vice versa. Can we take advantage of this fact? I think yes.

So, I propose the following modules (as a first approximation, for discussion):

1. `type-classes-no-deps` which are type-classes whose definitions do not require any dependencies on any other data-types. For example, `trait Functor`. The instances for core data types (e.g. `scala.List`) belong in `object Functor`.

2. `data-types-no-deps` which are data-types who definitions do not require any dependencies on any other type-classes. This is problematic, because it could be the introduction of a simple method to a data type, which changes that dependency. It would suck to have to care about introducing a method, simply for this reason. I leave this point open.

3. `type-classes-data-types-with-deps` which are both type-classes and data-types that have inter-dependencies. This is "everything else." Alternatively, the chain could keep going. For example, "type-classes that depend on data-types-with-no-deps" and vice versa, but eventually, it must be broken. I propose this as the optimal point.

Please discuss.

**Comment by aloiscochard** [comment]
:+1: about not using subtyping, I have been beaten so many times by it that I don't see any benefit to use it.

About the modules split, I have played a bit with those ideas on an [experimental branch][] and I think that 2. `data-types-no-deps` is unnecessary.

I would love counter examples (data type that have no instances), if anyone can think of any?

Also in my experiment I have a `prelude` module who is basically added as dependency to all other modules. It contains the necessary stuff to replace the default `Predef`, as well as the `Either` and `Maybe` type.

The reason is that those types seems quite fundamental to the definition of classes, by having `Either` in `prelude` I was able to keep `Profunctor` with `Choice` in the `class` module.

[experimental branch]: https://github.com/scalaz/scalaz/tree/series/8.0.x

**Comment by puffnfresh** [comment]
Maybe I'm misunderstanding something. The goal is to separate type-classes from data structures?

What do we do about things like this?

```scala
package scalaz

trait Foldable[F[_]] {
  def toList[A](fa: F[A]): scalaz.IList[A]
}
```

I don't want that to return a `scala.List`, that's for sure. There's going to be similar problems with `ProChoice` presenting `scalaz.\/` vs `scala.Either`. I'm sure there's a lot more.

I also don't see this separation as helping with modularity. Hopefully I'm misunderstanding.

**Comment by xuwei-k** [comment]
There was similar discussion one years ago.

https://groups.google.com/d/topic/scalaz/duLCHveSjmg/discussion

@retronym said

> I considered a more fine grained scalaz-core in the first round of modularization, but quickly came to the conclusion that it wouldn't work: type class implementations use data structures and Scala stdlib typeclass instances. Trying to cut all of those cords leads to anaemic APIs and copy/paste in implementations.

**Comment by aloiscochard** [comment]
@puffnfresh that is exactly why I created a `prelude` module in my prototype, so IList (or anything that in fact "fix" datastructure of the base library.

I had to do that because I ended in the exact situation you describe for `Profunctor` with `scala.Either` vs `scalaz.Either`.

So currently in the prototype:
- `prelude`: contain a replacement for Predef, with as well replacement for non-principled datastrucutre of the std-lib (think Maybe/Either/List/...)
- `class`: which depend only on `prelude`
- `core`: which depend on `prelude` and `class`

That seems to work fine so far, and this mean that a project could potentially depend just on `class` and `prelude` but not core.

If we realize later that this design is not realistic, we can always fuse `core` and `class`. But so far that seems okay...

The prototype is here:
https://github.com/scalaz/scalaz/tree/series/8.0.x

Also, @retronym mentioned somewhere else that he gave up on at the time on the no subtyping design due to problem when trying to put some implicit `Monad => Functor`.

I agree such implicits are a terrible idea, but there is an other way to work around the problem he faced, and it's like this:
https://github.com/aloiscochard/scalaz/commit/d300ff27b8b4d3ecbf3ba6a1a1be3427dcca74bc

I'll apply this on the prototype, and hopefully I should have a feeling of how realistic the approach is moving forward.

Feedback welcome!

**Comment by aloiscochard** [comment]
@puffnfresh BTW, just to be clear I'm not sure -as you said- if the separation does *really* bring much in term of modularity.

I have read a few time people complaining about scalaz not being modular enough, but I don't really get their point... when I use scalaz, I want the full stuff... of course, things like a RTS system, the machines libarry, ... big component like that should be separeted.

But I never really got the point about classes being separated, beside organisation of the code.

**Comment by puffnfresh** [comment]
@aloiscochard awesome! Your code looks great and I completely agree with all of your points.

**Comment by wedens** [comment]
imho separation on package level is enough

**Comment by aloiscochard** [comment]
@wedens yeah I think I agree for the `class` / `core` split.

I can still see `prelude` as being useful, specially in the light of having an `io` specification which could be decoupled and shared with other projects.

**Comment by puffnfresh** [comment]
Modularity improvements I care about:

* Transformers such as EitherT only requires constraints on `M[_]` that they actually need (i.e. not always Monad)
* Being able to refactor and not have code break (see the current mess with imports and ToFunctorOps)

**Comment by aloiscochard** [comment]
@puffnfresh thanks for sharing this.

1. sounds good to me, I have not investigate transformers encoding yet though.

2. my gut feeling is that this kind of issue should be solved with the new approach, I would love to be sure to fully understand the problem, do you have a pointer?

**Comment by puffnfresh** [comment]
Pretend I had a program where I imported Monad and then I wanted to import Traverse for some operations:

```scala
import scalaz.Maybe
import scalaz.syntax.monad._
import scalaz.syntax.traverse._

object Main {
  Maybe.just(1) as "problem"
}
```

This code use to work, but now I get:

```
/tmp/Example.scala:6: error: value as is not a member of scalaz.Maybe[Int]
  Maybe.just(1) as "problem"
                ^
one error found
```

The solution is:

```scala
import scalaz.Maybe
import scalaz.syntax.monad._
import scalaz.syntax.traverse.{ ToFunctorOps => _, _ }

object Main {
  Maybe.just(1) as "problem"  
}
```

You have to know how Scala is going to use an implicit to add methods. I think this really sucks.

**Comment by aloiscochard** [comment]
@puffnfresh very good point, I have faced multiple with foldable/traversable or apply/applicative.

It seems to me that the only solution for that problem is to prefix/suffix the `To*Ops` type with the name of type class it concern.

I'll try to think more about this, and apply that in the prototype. ty

**Comment by mpilquist** [comment]
@puffnfresh @aloiscochard Take a look at the code generation strategy I'm using in https://github.com/mpilquist/simulacrum. Specifically, how `ToXyzOps` and `ToAllXyzOps` are generated.

**Comment by aloiscochard** [comment]
@mpilquist nice ty! I was actually thinking yesterday, about something like `simulacrum` (or maybe an other "profile" in it), for end users to generate typeclass the way they are encoded in scalaz8.

I might get in touch with you once the design stabilize :)

**Comment by shajra** [comment]
> I think it can be concluded that subtyping in Scala is a failed experiment. The
> use of subtyping, even among type-classes, leads to problems, which I hope to
> discuss further in this issue. The downside to this is a code blowout.

I'd like to come back to this claim.  @puffnfresh recently showed me this [pull request against Dotty][PR] that allows ambiguous instances.  This happens more easily when using subtyping to encode type class hierarchies.  However, if we accept that the burden is on us to maintain coherence, then I wonder if this pull request sates our complaints with subtyping as an encoding.  I have a feeling @tonymorris will still be dead-set against subtyping (or not?).  But in light of this pull request, what about @S11001001 or @aloiscochard or anyone else?

@aloiscochard's "scato" approach for Scalaz 8 is a compelling design to get us off of subtype-encoding, but we still have more work to prove it out.  @puffnfresh has a concern about the `BaseHierarchy` abstraction, which he worries is not easy to extend with custom type classes.

So beyond trouble with ambiguous instances, what other complaints do we have with subtype encoding?  Also, it would be wonderful to have small code examples that illustrate complaints.  Otherwise, the arguments are often not grounded enough.

[PR]: https://github.com/lampepfl/dotty/pull/2046

**Comment by tonymorris** [comment]
I still believe subtyping is a failed experiment. I don't know how that PR demonstrates any further benefit.

This idea of, "but subtyping works if you do such and such" is a recurring theme, which is on its own, provokes deep suspicion.

**Comment by shajra** [comment]
@tonymorris subtyping definitely complicates implicit resolution.  Let me see if I can sum up your argument, and let me know if I get it wrong:

**My paraphrasing**:  Implicit resolution is already complicated enough.  Scalaz should seek simplicity, not complexity, and subtyping is a false economy of complexity.  We pay the price, and receive too few benefits.  If the "scato" approach gives us everything we need (especially if with not-to-much boilerplate), then we will have found the better architecture, and completely remove our reliance on odd compiler hacks like annotations to allow ambiguous instances.

Is that a fair assessment of your position?  In my eyes, this is sufficient reasoning to continue exploring "scato" even if we can make the other approach work.

I'm still interested in whether we know of any other problems beyond ambiguous instances -- something the pull request wouldn't solve.

Also, I want to say that even the "scato" approach doesn't completely remove dependence on subtyping, which is still used to establish precedence for implicit resolution.  It's just less subtyping, and ambiguous instances could still be an issue.

**Comment by tonymorris** [comment]
That's pretty close. In order to get subtyping working, you do a little bit of work for a penalty on the user side. The alternative is a tiny bit more work, to get a significant benefit.

I still think the best way to resolve this, is to come up with the smallest set of constraints, that is still diverse enough to demonstrate every issue or related matter, then write the code for it, including the user side, then ask where the benefits and penalties lie.

I have been done over enough times by scala's awful subtyping/implicit resolution mechanism, on the belief that it is slightly useful if you bend it the right way, that my position has become one more of ambivalence, until I see the code.

**Comment by shajra** [comment]
Okay, then I'm going to invest some time into exploring the "scato" approach.  If we don't, we will never know whether it works or not.  Also, we'll have concrete code to compare with a subtype-encoding approach.

On that note, I nearly have a small set of monad transformers working (ReaderT, EitherT, StateT, MonadReader, MonadError, MonadState, MonadTrans, including functions, instances, and syntax).  It's not complete by any means, but monad transformers are step more complex from the perspective of typeclass instance resolution.

Thus far, everything has been working rather nicely.  I'll report back when it's in a more presentable state.

**Comment by edmundnoble** [comment]
@shajra Because I know what your talk is about in a couple days, I have only this suggestion to make: remove the `Monad` constraint from `MonadReader`, `MonadError`, `MonadState` and friends. It is absolutely unnecessary and a mistake of the past; users need to demand the `Monad` constraint themselves or they are forced out of being able to abstract over `Applicative` and `Functor` with the same capabilities, and there is nothing intrinsic to `Reader`, `Error` and `State` that makes them "only work well" with the full powers of `Monad`. Also, my talk is more convincing if you go down that road ;)

**Comment by shajra** [comment]
@edmundnoble, you totally caught me off guard with that request.  I can do it, but I'm running out of time for a refactor (please forgive me if I don't get to it).

The names `Reader` and `State` are already taken as type aliases of their respective transformers applied to `Identity`.  What do you recommend the new names be?  If there's no `Monad` constraint, continuing to call them `MonadReader`, `MonadError`, and `MonadState` seems in poor taste.

**Comment by shajra** [comment]
I'm leaning towards the names `WithReader`, `WithError`, and `WithState`.  Any objections?  Going once...

**Comment by shajra** [comment]
@edmundnoble after some discussion with @tonymorris and @puffnfresh there's are strong argument to leave in the `Monad` constraint -- namely that laws we want to hold are defined in terms of `flatMap`.

This seems like a strong reason to leave the constraint in to me.

**Comment by puffnfresh** [comment]
Ideally for MonadState you'd instead have:

* ApplyState (e.g. law of `forall a b. put a *> put b = put b`)
* BindState (law of `forall f. get >>= \s -> get >>= f s = get >>= \s -> k s s`)
* MonadState (law of `get >>= put = return ()`)

**Comment by xuwei-k** [comment]
I have implemented `MonadState` laws in scalaprops based on [ClassLaws](https://hackage.haskell.org/package/ClassLaws-0.3.1.0/docs/Control-Monad-State-Class-Laws.html)

https://github.com/scalaprops/scalaprops/blob/v0.4.1/scalazlaws/src/main/scala/scalaprops/scalazlaws/MonadStateLaws.scala

**Comment by jdegoes** [comment]
@shajra Such laws can be expressed in terms of `MonadFree` with `liftF :: f a -> t f a`, and `foldFree :: (Monad g) => (f ~> g) -> Free f a -> g a` (or `MonadFreeAp`). It's more cumbersome but then the resulting type classes become usable in many more situations.

**Comment by edmundnoble** [comment]
@puffnfresh @tonymorris @shajra I do not understand how providing up to `n` (Apply, Bind, Monad, Applicative, Functor, ApplicativePlus, MonadCombine) x `m` (Reader, Writer, State, Either) combinations of typeclasses is practical; and until you *do* provide such, you are making a decision about *my* typeclass usage which forces me to use more operations than I need, making my abstractions larger than necessary and ironically making them harder to reason about. If you do not have the *power* to monadically inspect the state, *why* does it matter how monadically inspecting the state would work?

**Comment by edmundnoble** [comment]
@jdegoes very interesting idea; and generalizable to other Free structures.

**Comment by puffnfresh** [comment]
I don't think it's impractical to provide a more modular hierarchy for MTL. We should do that.

**Comment by edmundnoble** [comment]
@puffnfresh :+1: on that, for sure.

**Comment by fommil** [comment]
in response to [an old comment in this thread](https://github.com/scalaz/scalaz/issues/895#issuecomment-73546515) I've certainly noticed a lot of inconsistent use of `List` vs `IList` and I'd like scalaz8 (or even the next scalaz7) to standardise on one of them. In `Foldable` it's hard to know what you're going to get back sometimes, e.g. `splitBy` vs `splitWith`.

**Comment by jdegoes** [comment]
Agreed. I think 8 should only use safe and invariant structures, and I'd even go so far as to suggest dropping the `I` from the invariant structures (`IList` -> `List`) since you can always `import scala.collection.{List => CList}` or similar. If you're not using unsafe methods or esoteric collection methods, they should mostly be a drop-in replacement.

**Comment by tonymorris** [comment]
@fommil I agree, ditch `List` then provide a border API to interact with the legacy.

**Comment by fommil** [comment]
Hypelevel have proven that modularity buys nothing except user confusion and dependency hell.

The removal of subtyping has already made its way into scalaz 7.3 as we kill `+` and `-` type parameters everywhere we find them.

**Comment by tonymorris** [comment]
On the contrary, modularity buys you everything. Some attempted implementations of modularity do not achieve the goal :)

Modularity in scala requires certain trade-offs with how poorly-designed the language is. Failure to admit to those trades can often result in "worst of all worlds" scenarios.

**Comment by ritschwumm** [comment]
@jdegoes i think reusing a widely known and used (and imported by default...)  name List for anything else than the standard library is not a good idea at all. having to do the `import scala.collection.{List => CList}` might be annoying, but i could live with it.

a real problem this would be for reviewing git commits. there, not being able to distinguish the two Lists from the diff alone (with reasonable certainty) without looking at the imports of the whole file, too would increase cognitive load a lot.
